### ISSUE TITLE
[xray] Sending too much uncommitted lineage when forwarding tasks

### ISSUE THREAD

**Issue by stephanie-wang** [ray-project/ray#2530]
### Describe the problem

When forwarding tasks from one raylet to another, the message includes the uncommitted lineage of the forwarded task. However, in many cases, this may include many tasks that have already been sent in a previous task.

In particular, consider a workload where one raylet submits N methods to an actor on a second raylet, and nothing gets evicted yet from the lineage cache. We will end up spending O(N^2) on the `ForwardTask` messages, since each message contains all of the previously submitted actor methods.

We can reduce this to the optimal O(N) if we associate each task entry in the lineage cache with a hash set of the node managers that we've sent it to so far. We can then check this hash set as we walk a task's lineage in `GetUncommittedLineage`. Using this algorithm, each `ForwardTask` message in the workload above should contain only the forwarded task.

### Source code / logs
Adapted from a script provided in #2403. This test starts two raylets and submits 100,000 methods from one actor to another.
```
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import ray
import time
import datetime

@ray.remote(resources={"Node1": 1})
class A(object):
    def __init__(self):
        print("Actor A start...")

    def f(self):
        num_of_items = 100 * 1000

        b = B.remote(num_of_items)
        time_str = datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S:%f')
        print("push, start " + time_str)
        for i in range(num_of_items):
            b.f.remote(i)
        time_str = datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S:%f')
        print("push, end " + time_str)
        time.sleep(5)

@ray.remote(resources={"Node2": 1})
class B(object):
    def __init__(self, num_of_items):
        print("Actor B start...")
        self.num_of_items = num_of_items
        self.sum = 0

    def f(self, object_id):
        #ray.get(object_id)
        #print (object_id)
        self.sum += 1
        if (self.sum == 1) :
            time_str = datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S:%f')
            print("read, start " + time_str)

        if (self.sum == self.num_of_items) :
            time_str = datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S:%f')
            print("read, end " + time_str)
        if (self.sum % 10000 == 0) :
            time_str = datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S:%f')
            print("read " + str(self.sum) + "  " + str(time_str))


    def get_sum(self):
        return self.sum


if __name__ == "__main__":
    ray.worker._init(start_ray_local=True, use_raylet=True, num_local_schedulers=2, resources=[
        {
            "Node1": 1,
        },
        {
            "Node2": 1,
        }])
    A.remote().f.remote()
    time.sleep(300)
```

**Durations for the above test**
Legacy ray: 1min
xray: >3min
xray + #2438: >2min
xray + #2438 + artificially limiting the uncommitted lineage to size 1: 45s

**Comment by zhijunfu** [comment]
@stephanie-wang  Nice to see this! 
I was thinking about a slight different approach on the issue. For each task of an actor, its dependencies come from two parts: its arguments and its previous task. In this problem, there is a chain of uncommitted previous tasks from the same actor, which makes it very time consuming. When forwarding a task for an actor, maybe we could just forward the current task and its arguments, since we know its previous tasks have already been forwarded. It would probably require less  code changes, the downside is it would only apply to actors, while your approach is more generic.

**Comment by stephanie-wang** [comment]
@zhijunfu, yes, that is actually how we got the durations above, but I think in this case it's better to use a generic approach. That way, we don't have to determine which uncommitted lineage came from the arguments of an actor method versus the previously submitted methods.

**Comment by zhijunfu** [comment]
@stephanie-wang yes agreed:-)

**Comment by stephanie-wang** [comment]
Closed by #2534.
